cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamService.java

```diff
@@ -381,11 +377,25 @@ private void performLeaderElection() {
       zkClient, "/election/" + STREAMS_COORDINATOR, new ElectionHandler() {
       @Override
       public void leader() {
-        LOG.info("Became Stream handler leader. Starting resource coordinator.");
-        resourceCoordinator = new ResourceCoordinator(getCoordinatorZKClient(), discoveryServiceClient,
-                                                      new BalancedAssignmentStrategy());
-        resourceCoordinator.startAndWait();
-        updateRequirement();
+        while (true) {
+          LOG.info("Became Stream handler leader. Starting resource coordinator.");
+          heartbeatsSubscriptionExecutor = Executors.newSingleThreadExecutor(
+            Threads.createDaemonThreadFactory("heartbeats-subscription-executor"));
+          try {
+            heartbeatsSubscription = subscribeToHeartbeatsFeed();
+            resourceCoordinator = new ResourceCoordinator(getCoordinatorZKClient(), discoveryServiceClient,
```

Comment:
So you can actually move creation and starting of `ResourceCoordinator` outside of the loop, right?